gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/jupiter/reactor/handler/context/AbstractExecutionContext.java

```diff
@@ -186,20 +190,48 @@ public <T> Map<String, T> getInternalAttributes() {
     public TemplateEngine getTemplateEngine() {
         if (templateEngine == null) {
             templateEngine = TemplateEngine.templateEngine();
+            prepareTemplateEngine(templateEngine);
+        }
+
+        return templateEngine;
+    }
 
-            final TemplateContext templateContext = templateEngine.getTemplateContext();
-            final EvaluableRequest evaluableRequest = new EvaluableRequest(request());
-            final EvaluableResponse evaluableResponse = new EvaluableResponse(response());
+    @Override
+    public TemplateEngine getTemplateEngine(Message message) {
+        final TemplateEngine templateEngine = TemplateEngine.templateEngine();
 
-            templateContext.setVariable(HttpExecutionContext.TEMPLATE_ATTRIBUTE_REQUEST, evaluableRequest);
-            templateContext.setVariable(HttpExecutionContext.TEMPLATE_ATTRIBUTE_RESPONSE, evaluableResponse);
-            templateContext.setVariable(HttpExecutionContext.TEMPLATE_ATTRIBUTE_CONTEXT, new EvaluableExecutionContext(this));
+        prepareTemplateEngine(templateEngine);
 
-            if (templateVariableProviders != null) {
-                templateVariableProviders.forEach(templateVariableProvider -> templateVariableProvider.provide((ExecutionContext) this));
-            }
-        }
+        templateEngine.getTemplateContext().setVariable(TEMPLATE_ATTRIBUTE_MESSAGE, new EvaluableMessage(message));
 
         return templateEngine;
     }
+
+    private void prepareTemplateEngine(final TemplateEngine templateEngine) {
+        final TemplateContext templateContext = templateEngine.getTemplateContext();
+        final EvaluableRequest evaluableRequest = getEvaluableRequest();
+        final EvaluableResponse evaluableResponse = getEvaluableResponse();
+
+        templateContext.setVariable(HttpExecutionContext.TEMPLATE_ATTRIBUTE_REQUEST, evaluableRequest);
+        templateContext.setVariable(HttpExecutionContext.TEMPLATE_ATTRIBUTE_RESPONSE, evaluableResponse);
+        templateContext.setVariable(HttpExecutionContext.TEMPLATE_ATTRIBUTE_CONTEXT, new EvaluableExecutionContext(this));
```

Comment:
Shouldn't you do the same trick as `getEvaluableRequest` or `getEvaluableResponse`? I appreciate that current contexte could change but as we have stopped using wrapper we should be able to use object reference don't you think?

Reply:
`EvaluableRequest`, `EvaluableResponse` and `EvaluableMessage` wrappers are required for security reason to limit what can be accessed from EL expression. Also these wrappers don't replace the existing request or response like we had in the previous engine 😅

Reply:
Ok I was completely off 🤦🏻‍♂️. Makes sense, just forgot it 👍